Add `display_override` article (#4675)
GoogleChrome/web.dev: src/site/content/en/blog/display-override/index.md

---
layout: post
title: Preparing for the display modes of tomorrow
subhead: PWAs can use the "display_override" property to deal with special display modes.
authors:
  - thomassteiner
date: 2021-02-16
description: |

>   The display_override property allows developers to define a customized fallback chain of modes

Reviewer: ```suggestion
  The display_override property allows developers to define a customized fallback chain of modes.
```

Author: This sentence continues:

The display_override property allows developers to define a customized fallback chain of modes their PWAs should be displayed in.

  their PWAs should be displayed in.
hero: hero.jpg
alt: Web App Manifest source code excerpt.
tags:
  - blog # blog is a required tag for the article to show up in the blog.

>   - progressive-web-apps

Reviewer: What about creating a web app manifest tag? Do we have more than two articles?

Author: Added in https://github.com/GoogleChrome/web.dev/pull/4675/commits/1032982659136da400ff5eba2a032655ccccead2.

>   - progressive-web-apps

Author: ```suggestion
  - progressive-web-apps
  - web-app-manifest
```

---

A [Web App Manifest](/add-manifest/) is a JSON file that tells the browser about your Progressive
Web App and how it should behave when installed on the user's desktop or mobile device.
Via the [`display`](/add-manifest/#display) property, you can customize what browser UI is shown when your app is launched. For example, you can hide the
address bar and browser chrome. Games can even be made to launch full screen.

> As a quick recap, below are the display modes that are currently specified.

Reviewer: ```suggestion
As a quick recap, below are the display modes that are specified at the time this article was written.
```

Reviewer: I think your version is too easily interpreted as 'we check this regularly and keep it up to date.'

      </tr>
    </tbody>
  </table>
</div>

These display modes follow a [well-defined fallback chain](https://w3c.github.io/manifest/#dfn-fallback-display-mode)
(`"fullscreen"` → `"standalone"` → `"minimal-ui"` → `"browser"`). If a browser does not support a given
mode, it falls back to the next display mode in the chain.

## Shortcomings of the `display` property

The problem with this hard-wired fallback chain approach is threefold:

- A developer cannot request `"minimal-ui"` without being forced back into the `"browser"` display mode in case `"minimal-ui"` is not supported by a given browser.
- Developers have no way of handling cross-browser differences, like if the browser includes or excludes a [back button](https://twitter.com/ChromiumDev/status/1012065260625383425/photo/1) in the window for `"standalone"` mode.
- The current behavior makes it impossible to introduce new display
modes in a backward compatible way, since explorations like tabbed application mode do not have a
natural place in the fallback chain.

## The `display_override` property

{% Aside %}
  The `display_override` property is meant to solve special corner cases. In almost all

>   circumstances the regular `display` property is what developers should reach out for.

Reviewer: ```suggestion
  circumstances the regular `display` property is what developers should reach for.
```

{% endAside %}

> These problems are solved by the `display_override` property, which the browser considers *before*

Reviewer: >These problems

Your note interrupts a direct conceptual connection between the last section and this statement. If there were no such connection, I would have no objection to the placement of the node. Because of it, I think the aside should go after this paragraph.  Otherwise, I'm asking 'what problems' and I need to scroll up.

Author: Moved the note after the paragraph.

the `display` property. Its value is a sequence of strings that are considered in-order, and the
first supported display mode is applied. If none are supported, the browser falls back to evaluating

> the `display` field.

Author: ```suggestion
## The `display_override` property

These problems are solved by the `display_override` property, which the browser considers *before*
the `display` property. Its value is a sequence of strings that are considered in-order, and the
first supported display mode is applied. If none are supported, the browser falls back to evaluating
the `display` field.

{% Aside %}
  The `display_override` property is meant to solve special corner cases. In almost all
  circumstances the regular `display` property is what developers should reach out for.
{% endAside %}
```

> ## The `display_override` property
> 
> {% Aside %}
>   The `display_override` property is meant to solve special corner cases. In almost all
>   circumstances the regular `display` property is what developers should reach for.
> {% endAside %}
> 
> These problems are solved by the `display_override` property, which the browser considers *before*
> the `display` property. Its value is a sequence of strings that are considered in-order, and the
> first supported display mode is applied. If none are supported, the browser falls back to evaluating
> the `display` field.

Author: ```suggestion
## The `display_override` property

These problems are solved by the `display_override` property, which the browser considers *before*
the `display` property. Its value is a sequence of strings that are considered in-order, and the
first supported display mode is applied. If none are supported, the browser falls back to evaluating
the `display` field.

{% Aside %}
  The `display_override` property is meant to solve special corner cases. In almost all
  circumstances the regular `display` property is what developers should reach for.
{% endAside %}
```

In the example below, the display mode fallback chain would be as follows.

> (The details of `"window-control-overlay"` are out-of-scope for this article.)

Reviewer: That seems strange since it's one of the possible values for `display_override`.

Author: But one of potentially many. In the concrete case, I have an article coming at https://github.com/GoogleChrome/web.dev/pull/4680 that we can link to once it's live.

1. `"window-control-overlay"` (First, look at `display_override`.)
1. `"minimal-ui"`
1. `"standalone"` (When `display_override` is exhausted, evaluate `display`.)
1. `"minimal-ui"` (Finally, use the `display` fallback chain.)
1. `"browser"`

```json
{
  "display_override": ["window-control-overlay", "minimal-ui"],
  "display": "standalone",
}
```

{% Aside %}
  The browser will not consider `display_override` unless `display` is also present.
{% endAside %}

To remain backward compatible, any future display mode will only be acceptable as a value of
`display_override`, but not `display`.
Browsers that do not support `display_override` fall back to the `display` property and ignore
`display_override` as an unknown Web App Manifest property.

{% Aside %}

>   The `display_override` property is defined independent from its potential values.

Reviewer: ```suggestion
  The `display_override` property is defined independently from its potential values.
```

Reviewer: I'm not sure what this means, much less why it's relevant to me as a web developer. Does this mean that if my browser finds a `display_override` value it doesn't recognize that it will fail gracefully and roll back to something it does recognize?

Maybe that's what needs to be said.

Author: It's like standardizing the size of an envelop. We just say the envelop is 10x20x0.1. This makes no statement about what contents the envelop is designed for. It could be sheets of paper, leaves, or feathers; we don't care at this stage.

{% endAside %}

## Browser compatibility

The `display_override` property is supported as of Chromium&nbsp;89. Other browsers support the

> `display` property, which caters for the majority of display mode use cases.

Reviewer: ```suggestion
`display` property, which caters to the majority of display mode use cases.
```

---
layout: post
title: Preparing for the display modes of tomorrow
subhead: PWAs can use the "display_override" property to deal with special display modes.
authors:
  - thomassteiner

> date: 2021-02-16

Author: ```suggestion
date: 2021-02-25
```
